opm/output/Wells.hpp

```diff
@@ -0,0 +1,173 @@
+/*
+  Copyright 2016 Statoil ASA.
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef OPM_OUTPUT_WELLS_HPP
+#define OPM_OUTPUT_WELLS_HPP
+
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+namespace Opm {
+
+    namespace data {
+
+    class Rates {
+        /* Methods are defined inline for performance, as the actual *work* done
+         * is trivial, but somewhat frequent (typically once per time step per
+         * completion per well).
+         *
+         * To add a new rate type, add an entry in the enum with the correct
+         * shift, and if needed, increase the size type. Add a member variable
+         * and a new case in get_ref.
+         */
+
+        public:
+            using enum_size = uint16_t;
+
+            enum class opt : enum_size {
+                wat     = (1 << 0),
+                oil     = (1 << 1),
+                gas     = (1 << 2),
+                polymer = (1 << 3),
+            };
+
+            /// Query if a set of values are set. Returns true if all values
+            /// specified by the optmask are set, false if any are missing.
+            inline bool has( opt );
+
+            /// Read the value indicated by m. Throws an exception if multiple
+            /// values are requested or if the value is unset.
+            inline double get( opt m );
+            /// Read the value indicated by m. Returns errval if multiple
+            /// values are requested or the value is unset.
+            inline double get( opt m, double errval );
+            /// Set the value specified by m. Throws an exception if multiple
+            /// values are requested. Returns a self-reference to support
+            /// chaining.
+            inline Rates& set( opt m, double value );
+
+        private:
+            double& get_ref( opt, double& );
+
+            opt mask = static_cast< opt >( 0 );
+
+            double wat;
+            double oil;
+            double gas;
+            double polymer;
+    };
+
+    struct Completion {
+        size_t logical_cartesian_index;
+        Rates rates;
+    };
+
+    struct Well {
+        Rates rates;
+        double bhp;
+        std::map< size_t /*logical_cartesian_index*/, Completion > completions;
```

Comment:
Cute way to document what the map keys are, but I'd still prefer a normal, explanatory comment...